File: apps/backend/src/main/java/com/bsic/dataqualitybackend/service/RpaService.java
package com.bsic.dataqualitybackend.service;

import com.bsic.dataqualitybackend.model.RpaJob;
import com.bsic.dataqualitybackend.repository.RpaJobRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class RpaService {

    private final RpaJobRepository rpaJobRepository;
    private final WorkflowService workflowService;

    @Transactional
    public RpaJob createJob(Long ticketId, String processInstanceId, String action) {
        log.info("Creating RPA job for ticket: {}", ticketId);

        RpaJob job = RpaJob.builder()
                .jobId(UUID.randomUUID().toString())
                .ticketId(ticketId)
                .processInstanceId(processInstanceId)
                .action(action)
                .status("PENDING")
                .startedAt(LocalDateTime.now())
                .build();

        return rpaJobRepository.save(job);
    }

    @Transactional
    public void updateJobStatus(String jobId, String status, String errorMessage, String resultData) {
        log.info("Updating RPA job status: {} -> {}", jobId, status);

        RpaJob job = rpaJobRepository.findByJobId(jobId)
                .orElseThrow(() -> new IllegalArgumentException("RPA job not found: " + jobId));

        job.setStatus(status);

        if ("RUNNING".equals(status)) {
            job.setStartedAt(LocalDateTime.now());
        } else if ("COMPLETED".equals(status) || "FAILED".equals(status)) {
            job.setCompletedAt(LocalDateTime.now());
        }

        if (errorMessage != null) {
            job.setErrorMessage(errorMessage);
        }

        if (resultData != null) {
            job.setResultData(resultData);
        }

        rpaJobRepository.save(job);

        if ("COMPLETED".equals(status) || "FAILED".equals(status)) {
            notifyWorkflow(job);
        }
    }

    private void notifyWorkflow(RpaJob job) {
        if (job.getProcessInstanceId() != null) {
            boolean success = "COMPLETED".equals(job.getStatus());
            workflowService.notifyRpaCompletion(
                job.getProcessInstanceId(),
                success,
                job.getErrorMessage()
            );
        }
    }

    public Optional<RpaJob> getJobById(String jobId) {
        return rpaJobRepository.findByJobId(jobId);
    }

    public List<RpaJob> getJobsByTicket(Long ticketId) {
        return rpaJobRepository.findByTicketId(ticketId);
    }

    public List<RpaJob> getJobsByStatus(String status) {
        return rpaJobRepository.findByStatus(status);
    }

    @Transactional
    public void retryFailedJob(String jobId) {
        log.info("Retrying failed RPA job: {}", jobId);

        RpaJob job = rpaJobRepository.findByJobId(jobId)
                .orElseThrow(() -> new IllegalArgumentException("RPA job not found: " + jobId));

        if (!"FAILED".equals(job.getStatus())) {
            throw new IllegalStateException("Only failed jobs can be retried");
        }

        job.setStatus("PENDING");
        job.setRetryCount(job.getRetryCount() + 1);
        job.setStartedAt(LocalDateTime.now());
        job.setCompletedAt(null);

        rpaJobRepository.save(job);
    }

    public List<RpaJob> findStuckJobs(int timeoutMinutes) {
        LocalDateTime threshold = LocalDateTime.now().minusMinutes(timeoutMinutes);
        return rpaJobRepository.findStuckJobs(threshold);
    }

    @Transactional
    public void cleanupStuckJobs(int timeoutMinutes) {
        List<RpaJob> stuckJobs = findStuckJobs(timeoutMinutes);

        for (RpaJob job : stuckJobs) {
            log.warn("Marking stuck job as failed: {}", job.getJobId());
            updateJobStatus(job.getJobId(), "FAILED", "Job timeout - exceeded " + timeoutMinutes + " minutes", null);
        }
    }
}

File: apps/frontend/src/services/rpaService.ts
import { apiService } from './apiService';

export interface RpaJob {
  id: number;
  jobId: string;
  ticketId: number;
  processInstanceId?: string;
  action: string;
  status: 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED';
  startedAt?: string;
  completedAt?: string;
  errorMessage?: string;
  resultData?: string;
  retryCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface StartRpaJobRequest {
  ticketId: number;
  processInstanceId: string;
  action: string;
}

export interface RpaCallbackRequest {
  jobId: string;
  status: string;
  errorMessage?: string;
  resultData?: string;
}

class RpaService {
  private baseUrl = '/api/rpa';

  async startJob(request: StartRpaJobRequest): Promise<{ jobId: string }> {
    const response = await apiService.post<{ data: { jobId: string } }>(
      `${this.baseUrl}/jobs/start`,
      request
    );
    return response.data;
  }

  async getJob(jobId: string): Promise<RpaJob> {
    const response = await apiService.get<{ data: RpaJob }>(
      `${this.baseUrl}/jobs/${jobId}`
    );
    return response.data;
  }

  async getJobsByTicket(ticketId: number): Promise<RpaJob[]> {
    const response = await apiService.get<{ data: RpaJob[] }>(
      `${this.baseUrl}/jobs/ticket/${ticketId}`
    );
    return response.data;
  }

  async getJobsByStatus(status: string): Promise<RpaJob[]> {
    const response = await apiService.get<{ data: RpaJob[] }>(
      `${this.baseUrl}/jobs/status/${status}`
    );
    return response.data;
  }

  async retryJob(jobId: string): Promise<void> {
    await apiService.post(`${this.baseUrl}/jobs/${jobId}/retry`);
  }

  async getStuckJobs(timeoutMinutes: number = 30): Promise<RpaJob[]> {
    const response = await apiService.get<{ data: RpaJob[] }>(
      `${this.baseUrl}/jobs/stuck?timeoutMinutes=${timeoutMinutes}`
    );
    return response.data;
  }

  async cleanupStuckJobs(timeoutMinutes: number = 30): Promise<void> {
    await apiService.post(
      `${this.baseUrl}/jobs/cleanup-stuck?timeoutMinutes=${timeoutMinutes}`
    );
  }

  getStatusBadgeClass(status: RpaJob['status']): string {
    switch (status) {
      case 'COMPLETED':
        return 'bg-green-100 text-green-800';
      case 'RUNNING':
        return 'bg-blue-100 text-blue-800';
      case 'PENDING':
        return 'bg-yellow-100 text-yellow-800';
      case 'FAILED':
        return 'bg-red-100 text-red-800';
      default:
        return 'bg-gray-100 text-gray-800';
    }
  }
}

export const rpaService = new RpaService();
